src/protocol/vm/utils.rs

```diff
@@ -260,22 +267,24 @@ pub async fn get_code_for_contract(
 
     // Parse the address
     let addr: H160 = address.parse().map_err(|_| {
-        SimulationError::from(RpcError::InvalidRequest(format!(
-            "Failed to parse address: {}",
-            address
-        )))
+        SimulationError::RetryDifferentInput(format!("Failed to parse address: {}", address))
     })?;
 
     // Call eth_getCode to get the bytecode of the contract
     match provider.get_code(addr, None).await {
-        Ok(code) if code.is_empty() => Err(SimulationError::from(RpcError::EmptyResponse())),
+        Ok(code) if code.is_empty() => {
+            Err(SimulationError::RetryLater("Empty response from RPC".to_string()))
+        }
         Ok(code) => {
             let bytecode = Bytecode::new_raw(Bytes::from(code.to_vec()));
             Ok(bytecode)
         }
         Err(e) => {
             println!("Error fetching code for address {}: {:?}", address, e);
-            Err(SimulationError::from(RpcError::InvalidResponse(e)))
+            Err(SimulationError::RetryLater(format!(
+                "Invalid response from RPC: {:?}",
+                e.to_string()
```

Comment:
Here I think it strongly depends on the returned error... e.g. if we get a 404 then it should be fatal if we get sth like 500 it may be temporary..

Reply:
These are the provider errors that can be raised:

AFAIU only the `JsonRpcClientError` is a 500 so I'll make a case for that one. Let me know if I'm missing something else 

```
pub enum ProviderError {
    /// An internal error in the JSON RPC Client
    #[error("{0}")]
    JsonRpcClientError(Box<dyn crate::RpcError + Send + Sync>),

    /// An error during ENS name resolution
    #[error("ens name not found: {0}")]
    EnsError(String),

    /// Invalid reverse ENS name
    #[error("reverse ens name not pointing to itself: {0}")]
    EnsNotOwned(String),

    /// Error in underlying lib `serde_json`
    #[error(transparent)]
    SerdeJson(#[from] serde_json::Error),

    /// Error in underlying lib `hex`
    #[error(transparent)]
    HexError(#[from] hex::FromHexError),

    /// Error in underlying lib `reqwest`
    #[error(transparent)]
    HTTPError(#[from] reqwest::Error),

    /// Custom error from unknown source
    #[error("custom error: {0}")]
    CustomError(String),

    /// RPC method is not supported by this provider
    #[error("unsupported RPC")]
    UnsupportedRPC,

    /// Node is not supported by this provider
    #[error("unsupported node client")]
    UnsupportedNodeClient,

    /// Signer is not available to this provider.
    #[error("Attempted to sign a transaction with no available signer. Hint: did you mean to use a SignerMiddleware?")]
    SignerUnavailable,
}
```